eclair-core/src/main/scala/fr/acinq/eclair/payment/send/OfferPayment.scala

```diff
@@ -0,0 +1,157 @@
+import fr.acinq.eclair.{Features, InvoiceFeature, MilliSatoshi, NodeParams, TimestampSecond, randomKey}
+
+import java.util.UUID
+import scala.util.Random
+
+object OfferPayment {
+  sealed trait Failure
+
+  case class UnsupportedFeatures(features: Features[InvoiceFeature]) extends Failure
+
+  case class UnsupportedChains(chains: Seq[ByteVector32]) extends Failure
+
+  case class ExpiredOffer(expiryDate: TimestampSecond) extends Failure
+
+  case class QuantityTooHigh(quantityMax: Long) extends Failure
+
+  case class AmountInsufficient(amountNeeded: MilliSatoshi) extends Failure
+
+  case class InvalidSignature(signature: ByteVector64) extends Failure
+
+  case object NoInvoice extends Failure
+
+  case class InvalidInvoice(tlvs: TlvStream[OfferTypes.InvoiceTlv]) extends Failure
+
+  sealed trait Command
+
+  case class PayOffer(replyTo: ActorRef,
+                      offer: Offer,
+                      amount: MilliSatoshi,
+                      quantity: Long,
+                      sendPaymentConfig: SendPaymentConfig) extends Command
+
+  case class WrappedMessageResponse(response: OnionMessageResponse) extends Command
+
+  case class SendPaymentConfig(externalId_opt: Option[String],
+                               maxAttempts: Int,
+                               routeParams: RouteParams,
+                               blocking: Boolean)
+
+  def apply(nodeParams: NodeParams,
+            postman: typed.ActorRef[Postman.Command],
+            paymentInitiator: ActorRef): Behavior[Command] = {
+    Behaviors.setup(context =>
+      Behaviors.receiveMessagePartial {
+        case PayOffer(replyTo, offer, amount, quantity, sendPaymentConfig) =>
+          if (!nodeParams.features.bolt12Features().areSupported(offer.features)) {
+            replyTo ! UnsupportedFeatures(offer.features.invoiceFeatures())
+            Behaviors.stopped
+          } else if (!offer.chains.contains(nodeParams.chainHash)) {
+            replyTo ! UnsupportedChains(offer.chains)
+            Behaviors.stopped
+          } else if (offer.expiry.exists(_ < TimestampSecond.now())) {
+            replyTo ! ExpiredOffer(offer.expiry.get)
+            Behaviors.stopped
+          } else if (offer.quantityMax.getOrElse(1L) < quantity) {
+            replyTo ! QuantityTooHigh(offer.quantityMax.getOrElse(1))
+            Behaviors.stopped
+          } else if (offer.amount.exists(_ * quantity > amount)) {
+            replyTo ! AmountInsufficient(offer.amount.get * quantity)
+            Behaviors.stopped
+          } else {
+            val payerKey = randomKey()
+            val request = InvoiceRequest(offer, amount, quantity, nodeParams.features.bolt12Features(), payerKey, nodeParams.chainHash)
+            sendInvoiceRequest(nodeParams, postman, paymentInitiator, context, request, payerKey, replyTo, nodeParams.onionMessageConfig.maxAttempts, sendPaymentConfig)
+          }
+      })
+  }
+
+  val rand = new Random
```

Comment:
Instead of selecting a route at random, wouldn't it make more sense to iterate through them as we get failures, to ensure we try a different one? A simple way of doing that would be for example to replace `remainingAttempts` by `attemptNumber` (starting at 0) and using `blindedRoutes(attemptNumber % blindedRoutes.length)`.